src/validator/mod.rs

```diff
@@ -0,0 +1,143 @@
+use serde_json::{self, Value};
+
+use crate::{
+    schema::{PrimitiveType, Schema},
+    utils::value,
+};
+
+use scope::ScopedSchema;
+use state::ValidationState;
+
+mod error;
+mod path;
+mod scope;
+mod state;
+mod types;
+
+pub trait Validator {
+    fn validate(&self, data: Option<&Value>) -> ValidationState;
+}
+
+macro_rules! bail_if_invalid {
+    ($state:expr) => {{
+        let state = $state;
+        if !state.is_valid() {
+            return state;
+        }
+    }};
+}
+
+fn validate_optional(scope: &ScopedSchema, data: Option<&Value>) -> ValidationState {
+    let value_exists = match data {
+        Some(Value::Null) => false,
+        None => false,
+        _ => true,
+    };
+
+    if !value_exists && scope.schema().type_().is_required() {
+        return scope
+            .error(
+                "type",
+                format!("'{}' is not an optional type", scope.schema().type_().to_string()),
+            )
+            .into();
+    }
+
+    ValidationState::new()
+}
+
+fn validate_const(scope: &ScopedSchema, data: &Value) -> ValidationState {
+    match scope.schema().const_() {
+        Some(constant) if value::ne(constant, data) => scope.error("const", "value does not match").into(),
+        _ => ValidationState::new(),
+    }
+}
+
+fn validate_enum(scope: &ScopedSchema, data: &Value) -> ValidationState {
+    let enum_entries = scope.schema().enum_();
+
+    if enum_entries.is_empty() {
+        return ValidationState::new();
+    }
+
+    let valid_count = enum_entries
+        .iter()
+        .fold(0, |acc, item| acc + value::eq(item.value(), data) as usize);
+
+    match valid_count {
+        1 => ValidationState::new(),
+        0 => scope.error("enum", "does not match any value").into(),
+        _ => scope.error("enum", "matches multiple values").into(),
+    }
+}
+
+impl<'a> Validator for ScopedSchema<'a> {
```

Comment:
Could be moved to the top of the file maybe ? As this seems to be the entry point to the code in this module

Reply:
Same comment as in the previous one:

Don't have a strong opinion here. I prefer module flow in a way that if I have `X`, `Y` and `X` depends / uses `Y`, then the `Y` is first and then `X` (doesn't matter if it's a `pub` or not).